### DISCUSSION TITLE
에이전트별로 많은 양의 파라미터를 보낼 수 있는 방법이 있나요? Is there any way to send large amount of parameters per agent?

### DISCUSSION THREAD

**Discussion by KANG-JI-HUN** [naver/ngrinder#874] [Q&A] [answered]
(1) 현재 관제PC 1대 / 에이전트 20대를 세팅 후 스크립트 테스트를 진행해보고 있는데,
관제PC의 TXT또는 CSV파일에 들어있는 파라미터를 각 에이전트별로 보낼 수 있는 방법이 있는지 궁금합니다.

1~1000명까지의 파라미터를 보내려고 할 때,
**NeoLoad 성능 테스트 툴**에서는 관제PC의 CSV파일을 긁어와
---------------------------------------------------------------------------------------------------------------
에이전트1...    -> 1 ... 50
에이전트2...    -> 51 ... 100
에이전트3...    -> 101 ... 150
...
에이전트20...  -> 951 ... 1000
---------------------------------------------------------------------------------------------------------------
처럼 각 에이전트별로 파라미터를 나누어 부하를 주는 것으로 알고 있습니다.

**nGrinder 성능 테스트 툴**에서 스크립트를 작성할 때,
관제PC의 TXT or CSV파일을 긁어와 각 에이전트별로 파라미터를 보내는 방법이 있는지 궁금합니다.

(2) 만약 이러한 방법이 없다면, 최후의 방법으로 에이전트별로 TXT나 CSV파일을 넣어
관제PC의 스크립트를 통해 부하 테스트를 수행하고 싶은데 관련 예제가 있을까요?

ex) 각 에이전트별로 TXT or CSV파일을 넣어 관제PC의 스크립트를 수행
---------------------------------------------------------------------------------------------------------------
에이전트1...    -> 1 ... 50        (1번 텍스트 파일)
에이전트2...    -> 51 ... 100      (2번 텍스트 파일)
에이전트3...    -> 101 ... 150    (3번 텍스트 파일)
...
에이전트20...  -> 951 ... 1000  (20번 텍스트 파일)

**Comment by imbyungjun** [comment, marked as answer]
모든 에이전트에 동일하게 `1 ... 1000` 까지의 모든 파라미터 데이터를 갖는 파일을 리소스로 갖게합니다.
성능 테스트 스크립트 내에서 `grinder.agentNumber`를 호출하면 에이전트 마다 고유한 값을 얻을 수 있습니다. 만약 20개의 에이전트라면 각각 0~19 사이의 값을 얻게됩니다.
이 에이전트당 고유하게 부여받은 숫자를 기준으로 `1 ... 1000`을 나누어서 처리하도록 작성하면 될 것 같습니다.

**Reply by KANG-JI-HUN** [reply]
먼저 답변주셔서 감사합니다.
그럼 여기서 궁금한 점이 한가지 더 생기는데요,
각 에이전트 별로 고유한 값을 얻어 나누어서 처리를 하게 되면, 스크립트단에서 일종의 Ramp-Up을 거치게 되는게 아닌가요?
for문을 돌려 각 값에 따른 조건을 타는 것으로 이해를 했는데, 이렇게 되면 미세하게 부하 테스트에 틈이 생기는 게 아닌가 싶어서요.
모든 에이전트에 동일한 시간에 부하를 주고 싶은건데 방법이 있는지 답변 부탁드립니다.

**Reply by imbyungjun** [reply]
nGrinder의 부하 발생 방식에서 모든 에이전트가 대기했다가 동시에 부하를 발생시킬 방법은 고안되어있지 않습니다. 이게 정말로 중요한 테스트라면 같은 시점에 부하를 발생시킨다는것을 보장하도록 직접 구현을 해주셔야할 것 같습니다.
nGrinder의 에이전트는 부하 테스트가 시작되면 목표로 설정한 VUser에 도달할 때 까지 에이전트의 프로세스와 스레드를 생성하는데, 이 때 각각의 스레드는 준비가 되자마자 부하를 발생시키기 시작합니다. 성능 테스트 결과 그래프 중 TPS 지표가 초기에 0에서부터 점차적으로 평균 TPS까지 도달하게 되는것도 이 이유입니다.

말씀주신대로 미세한 부하 발생 시점의 차이는 어쩔 수 없지만, for문이 싫으시면 요런 형태로도 되지 않을까용??
```kt
val agentNumber = 0 // grinder.agentNumber를 통해 0~19 사이의 값을 각각 부여받음
val params = (1..1000).toList()
val agents = (1..20).toList()
val paramsPerAgent = params.chunked(params.size / agents.size).get(agentNumber)
println("${agentNumber}번 에이전트는 ${paramsPerAgent.first()}..${paramsPerAgent.last()} 파라미터를 처리한다.")
```

**Reply by KANG-JI-HUN** [reply]
감사합니다!
답변주신 내용으로 진행해보겠습니다~
